File: lib/models/chat_model.dart
class ChatMessage {
  final String id;
  final String senderId;
  final String text;
  final DateTime? timestamp;
  final bool isRead;

  ChatMessage({
    required this.id,
    required this.senderId,
    required this.text,
    this.timestamp,
    this.isRead = false,
  });

  factory ChatMessage.fromMap(Map<String, dynamic> map) {
    return ChatMessage(
      id: map['id'] ?? '',
      senderId: map['senderId'] ?? '',
      text: map['text'] ?? '',
      timestamp: map['timestamp'] != null
          ? DateTime.fromMillisecondsSinceEpoch(map['timestamp'] is int ? map['timestamp'] : (map['timestamp'] as double).toInt())
          : null,
      isRead: map['isRead'] ?? false,
    );
  }
}

class ChatConversation {
  final String chatId;
  final String otherUserId;
  final String otherUserName;
  final String? otherUserPhoto;
  final String lastMessage;
  final int lastUpdated;

  ChatConversation({
    required this.chatId,
    required this.otherUserId,
    required this.otherUserName,
    this.otherUserPhoto,
    required this.lastMessage,
    required this.lastUpdated,
  });

  factory ChatConversation.fromMap(Map<String, dynamic> map) {
    return ChatConversation(
      chatId: map['chatId'] ?? '',
      otherUserId: map['otherUserId'] ?? '',
      otherUserName: map['otherUserName'] ?? 'User',
      otherUserPhoto: map['otherUserPhoto'],
      lastMessage: map['lastMessage'] ?? '',
      lastUpdated: map['lastUpdated'] ?? 0,
    );
  }
}

File: src/models/chatModel.js
const { dbGet, dbSet, dbUpdate, dbGetAll, dbPush } = require('../config/firebase');

const CHATS_PATH = 'chats';
const BOOKINGS_PATH = 'bookings';

const ChatModel = {
  getChatId(uid1, uid2) {
    return [uid1, uid2].sort().join('_');
  },

  async getOrCreate(uid1, uid2) {
    const chatId = this.getChatId(uid1, uid2);
    const existing = await dbGet(`${CHATS_PATH}/${chatId}`);
    if (existing) return existing;

    const sorted = [uid1, uid2].sort();
    const chat = {
      chatId,
      participant1: sorted[0],
      participant2: sorted[1],
      participants: sorted,
      lastMessage: '',
      lastUpdated: Date.now(),
      _createdAt: Date.now(),
    };
    await dbSet(`${CHATS_PATH}/${chatId}`, chat);
    return chat;
  },

  /**
   * Get messages (latest N messages from RTDB).
   */
  async getMessages(chatId, limit = 50) {
    const { db } = require('../config/firebase');
    const snap = await db.ref(`${CHATS_PATH}/${chatId}/messages`)
      .orderByChild('timestamp')
      .limitToLast(limit)
      .once('value');

    if (!snap.exists()) return [];
    const messages = [];
    snap.forEach(child => {
      const value = child.val();
      messages.push({
        id: child.key,
        senderId: value.senderId || '',
        text: value.text || '',
        timestamp: value.timestamp || 0,
      });
    });
    return messages.sort((a, b) => a.timestamp - b.timestamp);
  },

  /**
   * Send a message.
   */
  async sendMessage(chatId, senderId, text) {
    const { db } = require('../config/firebase');
    const msgRef = db.ref(`${CHATS_PATH}/${chatId}/messages`).push();
    const message = {
      id: msgRef.key,
      senderId,
      text,
      timestamp: Date.now(),
    };
    await msgRef.set(message);

    // Update chat metadata for shared conversation
    await dbUpdate(`${CHATS_PATH}/${chatId}`, {
      lastMessage: text.length > 60 ? text.substring(0, 60) + '...' : text,
      lastUpdated: Date.now(),
    });

    return message;
  },

  /**
   * Check if two users have an active confirmed booking (access control for chat).
   */
  async hasActiveBooking(uid1, uid2) {
    const allBookings = await dbGetAll(BOOKINGS_PATH);
    return allBookings.some(b =>
      ['confirmed', 'in_progress', 'completed'].includes(b.status) &&
      ((b.customerId === uid1 && b.professionalId === uid2) ||
       (b.customerId === uid2 && b.professionalId === uid1))
    );
  },

  /**
   * Get all conversations for a user.
   */
  async getConversationsForUser(uid) {
    const all = await dbGetAll(CHATS_PATH);
    return all.filter(chat => {
      const participants = Array.isArray(chat.participants) ? chat.participants : Object.values(chat.participants || {});
      return participants.includes(uid);
    }).sort((a, b) => (b.lastUpdated || 0) - (a.lastUpdated || 0));
  },
};

module.exports = ChatModel;
